README.rst
Minform
=======

.. image:: https://img.shields.io/travis/daviddonna/minform.svg
   :target: https://travis-ci.org/daviddonna/minform
.. image:: http://img.shields.io/pypi/v/minform.svg
   :target: https://pypi.python.org/pypi/minform

The declarative ease of
`WTForms <https://github.com/wtforms/wtforms>`_, the small data
footprint of
`struct <https://docs.python.org/3/library/struct.html>`_.

For more detailed examples, read the full docs at
https://minform.readthedocs.org.

Why does Minform exist?
-----------------------

Let's talk about data size.

Say you have a fleet of devices with cell modems that report to a web server.
Here's an example of a data packet struct you might use:

.. code-block:: c

    struct sensor_data {
        char version[8];
        char serial[12];
        int32_t latitude;    // fixed-point * 100000
        int32_t longitude;   // fixed-point * 100000
        int16_t temperature; // fixed-point * 100
        uint16_t pings;
        uint8_t battery_pct;
    };

Let's say they're reporting their data in JSON format, because that's lighter
than XML but still coherent to most server frameworks. Here's a data packet:

.. code-block:: javascript

    {
        "version": "1.0",
        "serial": "DEADBEEF",
        "latitude": 4071270,
        "longitude": -7400590,
        "temperature": 3200,
        "pings": 123,
        "battery_pct": 62
    }

If you take out all the whitespace, that's **144 bytes**. (ASCII encoding).
Maybe that's all you need, but maybe you need to store billions or trillions
of these little guys. Worse, maybe you need to pay through the tear ducts for
cellular data. It would be nice for that data to be smaller, and to be
predictably sized.

Besides, let's face it: serializing to JSON from C is a pain in the patoot.
Depending on the library, you could be looking at ten or twenty lines of code
(or a truly epic ``sprintf``), just to serialize a record.

On the Python side, we can probably just use a form library like WTForms to
validate incoming data, but we've already paid a price to make that data
server-friendly.

What can Minform do for me?
---------------------------

Let's build a Minform form to handle incoming sensor data.

.. code-block:: python

    from minform import *

    class SensorData(BinaryForm):

        order = LITTLE_ENDIAN  # let's say our devices are little-endian
        version = BytesField(max_length=8, length=AUTOMATIC)
        serial = BytesField(max_length=12, length=AUTOMATIC)
        latitude = Int32Field()
        longitude = Int32Field()
        temperature = Int16Field()
        pings = UInt16Field()
        battery_pct = UInt8Field()

        # We could cut the serialized record off three bytes early, but let's
        # say we want to break on four-byte word boundaries for reasons.
        padding = BlankBytes(3)

Here's the C code that will serialize your structure:

.. code-block:: c

    #include <string.h>

    char send_buffer[36];
    struct sensor_data data;

    // ...

    memcpy(send_buffer, &data, sizeof(data));

And here's the Python that will receive it:

.. code-block:: python

    form = SensorData.unpack(serialized_data)

That serialized record is **36 bytes**. 36 on the wire, 36 in a file. You may
need to tweak the form definition, depending on your C compiler and the target
architecture, but Minforms gives you the tools to cope with padding bytes, and
even mixed byte ordering.

Let's fill in some gaps
-----------------------

Minforms are an awful lot like WTForms: you subclass ``minform.BinaryForm``,
and add ``BinaryField``\ s as class properties. Here's another quick example:

.. code:: python

    import minform

    class MyForm(minform.BinaryForm):
        '''
        This is a subclass of wtforms.Form: you can validate data with it,
        construct it from an HTML form, extract the data as a Python dict, etc.
        '''
        first_name = minform.BytesField('First Name', max_length=10)
        last_name = minform.BytesField('Last Name', max_length=10)
        age = minform.UInt8Field('Age')

    #               first_name (10)          last_name (10)           age (1)
    packed_data = b'David\x00\x00\x00\x00\x00Donna\x00\x00\x00\x00\x00\x18'
    form = MyForm.unpack(packed_data)
    assert form.data == {
        'first_name': b'David',
        'last_name': b'Donna',
        'age': 24,
    }

    next_form = MyForm(first_name=b'Foo', last_name=b'Barsson', age=100)
    packed = next_form.pack()
    assert packed == b'Foo\x00\x00\x00\x00\x00\x00\x00Barsson\x00\x00\x00\x64'

Because the library is built on ``struct``, binary serializations of a form's
data are in fixed-length buffers. This makes them easier to store, and easy to
map onto relatively naive serializations of C structs. It also allows for
clear documentation of the binary format, because the data maps predictably
onto different positions in a packed buffer.

Compound BinaryFields allow you to create nested structures that still
serialize into flat buffers.

.. code:: python

    class MyBigBadForm(minform.BinaryForm):
        """
        This is taking a turn for campy criminality.
        """
        riches = minform.Int16Field()
        goons = minform.BinaryFieldList(Person, max_entries=4, length=minform.EXPLICIT)

    squad = MyBigBadForm(riches=55223, goons=[
        {'first_name': 'Joey', 'last_name': 'Schmoey', 'age': 32},
        {'first_name': 'Manny', 'last_name': 'The Man', 'age': 40},
        {'first_name': 'Gerta', 'last_name': 'Goethe', 'age': 52},
    ])
    assert squad.pack() == (b'\xd7\xb7' +                                  # riches
                            b'\x03' +                                      # goons prefix
                            b'Joey\0\0\0\0\0\0Schmoey\0\0\0\x20' +         # goons[0]
                            b'Manny\0\0\0\0\0The Man\0\0\0\x28' +          # goons[1]
                            b'Gerta\0\0\0\0\0Goethe\0\0\0\0\x34' +         # goons[2]
                            b'\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0\0') # goons[3]

Even with an entire set of blank bytes for ``goons[3]``, that's 87 bytes, vs
185 for the JSON representation.


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (10/28/2015)

    Initial release

0.2.0 (11/22/2015)

    - Add pack_into and unpack_from methods to BinaryForm and BinaryItem.
    - Rename all 'buf' arguments to 'buffer'.

tests/util.py
import unittest
import pytest


class FormTest(unittest.TestCase):

    valid_inputs = []
    invalid_inputs = []

    def test_form_cannot_unpack_too_little_data(self):
        size = self.Form.size
        if size > 0:
            with pytest.raises(ValueError):
                self.Form.unpack(b'\x00' * (size - 1))

    def test_form_cannot_unpack_too_much_data(self):
        size = self.Form.size
        with pytest.raises(ValueError):
            self.Form.unpack(b'\x00' * (size + 1))

    def test_valid_inputs_are_accepted(self):
        for valid_input in self.valid_inputs:
            self.Form.unpack(valid_input)

    def test_invalid_inputs_are_not_accepted(self):
        for invalid_input in self.invalid_inputs:
            with pytest.raises(ValueError):
                self.Form.unpack(invalid_input)

    def test_form_can_unpack_from_exactly_sized_buffer(self):
        buffer = b'\0' * self.Form.size
        self.Form.unpack_from(buffer, 0)

    def test_form_cannot_unpack_from_small_buffer(self):
        buffer = b'\0' * (self.Form.size - 1)
        with pytest.raises(ValueError):
            self.Form.unpack_from(buffer, 0)

    def test_form_can_unpack_from_partway_along_buffer(self):
        buffer = b'\0' * (self.Form.size + 3)
        self.Form.unpack_from(buffer, 2)

    def test_form_can_unpack_from_exact_negative_index(self):
        buffer = b'\0' * self.Form.size
        self.Form.unpack_from(buffer, -self.Form.size)


tests/test_core.py
import pytest
import unittest
import wtforms
import minform


class TestBinaryItem(unittest.TestCase):

    class MyItem(minform.BinaryItem):

        size = 4

        def pack(self, data, order=None):
            return b'\x01\x02\x03\x04'

        def unpack(self, buf, order=None):
            return 0x01020304

    def test_binary_item_cannot_be_subclassed_without_pack(self):

        class MyItem(minform.BinaryItem):
            def unpack(self, buf, order=None):
                pass

        with pytest.raises(TypeError):
            MyItem()

    def test_binary_item_cannot_be_subclassed_without_unpack(self):

        class MyItem(minform.BinaryItem):
            def pack(self, data, order=None):
                pass

        with pytest.raises(TypeError):
            MyItem()

    def test_pack_into_works_for_same_size_buffer(self):
        m = self.MyItem()
        buf = bytearray(4)
        m.pack_into(buf, 0, None)
        assert buf == b'\x01\x02\x03\x04'

    def test_pack_into_fails_for_small_buffer(self):
        m = self.MyItem()
        buf = bytearray(3)
        with pytest.raises(ValueError):
            m.pack_into(buf, 0, None)

    def test_pack_into_fails_for_small_remaining_buffer(self):
        m = self.MyItem()
        buf = bytearray(5)
        with pytest.raises(ValueError):
            m.pack_into(buf, 2, None)

    def test_pack_into_fails_with_small_buffer_from_end(self):
        m = self.MyItem()
        buf = bytearray(5)
        with pytest.raises(ValueError):
            m.pack_into(buf, -3, None)

    def test_pack_into_succeeds_with_sufficient_buffer_from_end(self):
        m = self.MyItem()
        buf = bytearray(5)
        m.pack_into(buf, -4, None)

    def test_pack_into_succeeds_with_large_buffer_from_end(self):
        m = self.MyItem()
        buf = bytearray(5)
        m.pack_into(buf, -5, None)

    def test_unpack_from_works_for_same_size_buffer(self):
        m = self.MyItem()
        buf = b'\x01\x02\x03\x04'
        data = m.unpack_from(buf, 0, None)
        assert data == 0x01020304

    def test_unpack_from_fails_for_small_buffer(self):
        m = self.MyItem()
        buf = bytearray(3)
        with pytest.raises(ValueError):
            m.unpack_from(buf, 0, None)

    def test_unpack_from_fails_for_small_remaining_buffer(self):
        m = self.MyItem()
        buf = bytearray(5)
        with pytest.raises(ValueError):
            m.unpack_from(buf, 2, None)

    def test_unpack_from_fails_with_small_buffer_from_end(self):
        m = self.MyItem()
        buf = bytearray(5)
        with pytest.raises(ValueError):
            m.unpack_from(buf, -3, None)

    def test_unpack_from_succeeds_with_sufficient_buffer_from_end(self):
        m = self.MyItem()
        buf = bytearray(5)
        data = m.unpack_from(buf, -4, None)
        assert data == 0x01020304


class TestBinaryForm(unittest.TestCase):

    class Form(minform.BinaryForm):

        order = minform.BIG_ENDIAN

        char = minform.CharField()
        int32 = minform.Int32Field()
        bytes = minform.BytesField(max_length=6, length=minform.EXPLICIT)
        _ = minform.BlankBytes(3)
        lst = minform.BinaryFieldList(minform.UInt8Field(), max_entries=3,
                                      length=minform.FIXED)
        end = minform.BytesField(max_length=2, length=minform.AUTOMATIC)
        lst2 = minform.BinaryFieldList(minform.UInt16Field(), max_entries=1,
                                       length=minform.EXPLICIT)

    data = {
        'char': b'\x10',
        'int32': 0x12345678,
        'bytes': b'foo',
        'lst': [
            0x01,
            0x02,
            0x03,
        ],
        'end': b'ab',
        'lst2': [0x1234]
    }

    buf = (b'\x10\x12\x34\x56\x78' +
           b'\x03foo\0\0\0' +
           b'\0\0\0\x01\x02\x03' +
           b'ab\x01\x12\x34')

    size = len(buf)

    def test_non_field_binary_items_are_removed_from_namespace(self):
        assert not hasattr(self.Form, '_')

    def test_binary_items_are_replace_with_fields(self):
        for name in 'char int32 bytes lst end lst2'.split():
            assert isinstance(getattr(self.Form, name),
                              wtforms.core.UnboundField)

    def test_binary_items_are_cached_in_list(self):
        assert len(self.Form._binary_items) == 7
        assert all(isinstance(f, minform.BinaryItem)
                   for f in self.Form._binary_items)

    def test_pack_into_works_for_same_size_buffer(self):
        form = self.Form(data=self.data)
        buf = bytearray(self.size)
        form.pack_into(buf, 0)
        assert buf == self.buf

    def test_pack_into_fails_for_small_buffer(self):
        m = self.Form(data=self.data)
        buf = bytearray(self.size - 1)
        with pytest.raises(ValueError):
            m.pack_into(buf, 0)

    def test_pack_into_fails_for_small_remaining_buffer(self):
        m = self.Form(data=self.data)
        buf = bytearray(self.size + 1)
        with pytest.raises(ValueError):
            m.pack_into(buf, 2)

    def test_pack_into_fails_with_small_buffer_from_end(self):
        m = self.Form(data=self.data)
        buf = bytearray(self.size + 1)
        with pytest.raises(ValueError):
            m.pack_into(buf, -(self.size - 1))

    def test_pack_into_succeeds_with_sufficient_buffer_from_end(self):
        m = self.Form(data=self.data)
        buf = bytearray(self.size + 1)
        m.pack_into(buf, -self.size)

    def test_unpack_from_succeeds_for_same_size_buffer(self):
        form = self.Form.unpack_from(self.buf, 0)
        assert form.data == self.data

    def test_unpack_from_fails_for_small_buffer(self):
        with pytest.raises(ValueError):
            self.Form.unpack_from(self.buf[:-1], 0)

    def test_unpack_from_fails_for_small_remaining_buffer(self):
        buf = bytearray(self.size + 1)
        with pytest.raises(ValueError):
            self.Form.unpack_from(buf, 2)

    def test_unpack_from_fails_with_small_buffer_from_end(self):
        buf = bytearray(self.size + 1)
        with pytest.raises(ValueError):
            self.Form.unpack_from(buf, -(self.size - 1))

    def test_unpack_from_succeeds_with_sufficient_buffer_from_end(self):
        buf = b'\x00\x00' + self.buf
        form = self.Form.unpack_from(buf, -self.size)
        assert form.data == self.data


tests/test_basic.py
import pytest
import struct
import unittest
from wtforms.validators import NumberRange
import minform

from . import util

up_to_ten = b'\x00\x01\x02\x03\x04\x05\x06\x07\x08\x09'


class TestCharField(util.FormTest):

    class Form(minform.BinaryForm):
        char = minform.CharField()

    def test_form_has_field(self):
        assert len(self.Form._binary_items) == 1

    def test_char_field_can_be_loaded(self):
        form = self.Form(char='x')
        assert form.validate()
        assert form.data == {'char': 'x'}

    def test_char_field_refuses_empty_string(self):
        form = self.Form(char='')
        assert not form.validate()

    def test_char_field_refuses_long_string(self):
        form = self.Form(char='ab')
        assert not form.validate()

    def test_char_field_can_pack(self):
        form = self.Form(char=b'\x02')
        assert form.validate()
        b = form.pack()
        assert b == b'\x02'

    def test_char_field_can_unpack(self):
        form = self.Form.unpack(b'\x33')
        assert form.data == dict(char=b'\x33')


class IntField(object):

    def setUp(self):
        class Form(minform.BinaryForm):
            n = self.integer_field('n')
        self.Form = Form

    def test_too_small_number_is_rejected(self):
        n = self.integer_field.min - 1
        form = self.Form(n=n)
        assert not form.validate()

    def test_too_large_number_is_rejected(self):
        n = self.integer_field.max + 1
        form = self.Form(n=n)
        assert not form.validate()

    def test_min_number_is_accepted(self):
        n = self.integer_field.min
        form = self.Form(n=n)
        assert form.validate()

    def test_max_number_is_accepted(self):
        n = self.integer_field.max
        form = self.Form(n=n)
        assert form.validate()

    def test_zero_is_accepted(self):
        if 0 >= self.integer_field.min and 0 <= self.integer_field.max:
            form = self.Form(n=0)
            assert form.validate()

    def test_min_number_can_pack(self):
        n = self.integer_field.min
        form = self.Form(n=n)
        buf = form.pack()
        assert len(buf) == self.integer_field().size

    def test_max_number_can_pack(self):
        n = self.integer_field.max
        form = self.Form(n=n)
        buf = form.pack()
        assert len(buf) == self.integer_field().size

    def test_explicit_little_endian_deserialization(self):
        temp = self.integer_field()
        big_buf = b'\x00\x01\x02\x03\x04\x05\x06\x07'[:temp.size]
        little_value = struct.unpack('<' + temp.pack_string, big_buf)[0]
        form = self.Form.unpack(big_buf, order=minform.LITTLE_ENDIAN)
        assert form.data == dict(n=little_value)

    def test_implicit_little_endian_deserialization(self):
        temp = self.integer_field(order=minform.LITTLE_ENDIAN)

        class Form(minform.BinaryForm):
            n = self.integer_field(order=minform.LITTLE_ENDIAN)

        buf = b'\x00\x01\x02\x03\x04\x05\x06\x07'[:temp.size]
        little_value = struct.unpack('<' + temp.pack_string, buf)[0]
        form = Form.unpack(buf)
        assert form.data == dict(n=little_value)


class TestInt8Field(IntField, util.FormTest):
    integer_field = minform.Int8Field

    def test_extra_validators_are_passed_in(self):

        class Form(minform.BinaryForm):
            n = minform.UInt8Field(validators=[NumberRange(max=10)])

        form = Form(n=-1)
        assert not form.validate()

        form2 = Form(n=11)
        assert not form2.validate()


class TestUInt8Field(IntField, util.FormTest):
    integer_field = minform.UInt8Field


class TestInt16Field(IntField, util.FormTest):
    integer_field = minform.Int16Field


class TestUInt16Field(IntField, util.FormTest):
    integer_field = minform.UInt16Field


class TestInt32Field(IntField, util.FormTest):
    integer_field = minform.Int32Field


class TestUInt32Field(IntField, util.FormTest):
    integer_field = minform.UInt32Field


class TestInt64Field(IntField, util.FormTest):
    integer_field = minform.Int64Field


class TestUInt64Field(IntField, util.FormTest):
    integer_field = minform.UInt64Field


class TestFloat32Field(util.FormTest):

    class Form(minform.BinaryForm):
        f = minform.Float32Field()


class TestFloat64Field(util.FormTest):

    class Form(minform.BinaryForm):
        f = minform.Float64Field()


class TestBinaryBooleanField(util.FormTest):

    class Form(minform.BinaryForm):
        b = minform.BinaryBooleanField()

    def test_can_load_with_true(self):
        form = self.Form(b=True)
        assert form.validate()

    def test_can_load_with_false(self):
        form = self.Form(b=False)
        assert form.validate()

    def test_can_unpack_true(self):
        form = self.Form.unpack(b'\x01')
        assert form.data == dict(b=True)

    def test_can_unpack_false(self):
        form = self.Form.unpack(b'\x00')
        assert form.data == dict(b=False)

    def test_can_pack_true(self):
        form = self.Form(b=True)
        assert form.pack() == b'\x01'

    def test_can_pack_false(self):
        form = self.Form(b=False)
        assert form.pack() == b'\x00'


class TestFixedBytesField(unittest.TestCase):

    class Form(minform.BinaryForm):
        ten = minform.BytesField(max_length=10, length=minform.FIXED)
        empty = minform.BytesField(max_length=0)

    def check(self, buf):
        form = self.Form.unpack(buf)
        assert form.data == dict(ten=buf, empty=b'')
        assert form.pack() == buf

    def test_bytes_field_needs_max_length(self):
        with pytest.raises(ValueError):

            class Form(minform.BinaryForm):
                s = minform.BytesField(label='foo')

    def test_empty_field_has_width_0(self):
        assert self.Form.size == 10

    def test_fields_load_together(self):
        buf = up_to_ten
        form = self.Form.unpack(buf)
        assert form.data == dict(ten=buf, empty=b'')

    def test_terminal_null_byte(self):
        self.check(b'123456789\x00')

    def test_null_string(self):
        self.check(b'\x00' * 10)


class TestExplicitBytesField(unittest.TestCase):

    class ShortForm(minform.BinaryForm):
        s = minform.BytesField(max_length=10, length=minform.EXPLICIT)

    class LongForm(minform.BinaryForm):
        s = minform.BytesField(max_length=256, length=minform.EXPLICIT)

    def test_empty_field_has_width_zero(self):
        assert self.ShortForm.size == 11

    def test_fields_load_together(self):
        buf = b'\x0A' + up_to_ten
        form = self.ShortForm.unpack(buf)
        assert form.data == dict(s=up_to_ten)

    def test_read_stops_at_limit(self):
        buf = b'\x05' + up_to_ten
        form = self.ShortForm.unpack(buf)
        assert form.data == dict(s=b'\x00\x01\x02\x03\x04')

    def test_read_can_be_zero(self):
        buf = b'\x00' + up_to_ten
        form = self.ShortForm.unpack(buf)
        assert form.data == dict(s=b'')

    def test_larger_buffer_has_larger_count(self):
        assert self.LongForm.size == 258

    def test_invalid_count_is_flagged(self):
        with pytest.raises(ValueError):
            buf = b'\x0B' + up_to_ten
            self.ShortForm.unpack(buf)

    def test_bytes_field_packs_correctly(self):
        f = self.ShortForm(s=b'foo')
        buf = f.pack()
        assert buf == b'\x03foo\x00\x00\x00\x00\x00\x00\x00'


class TestVariableBytesField(util.FormTest):

    class Form(minform.BinaryForm):
        s = minform.BytesField(max_length=10, length=minform.AUTOMATIC)

    def test_full_bufer_stops_naturally(self):
        buf = up_to_ten
        form = self.Form.unpack(buf)
        assert form.data == dict(s=buf)

    def test_stops_before_trailing_null_bytes(self):
        bufs = [
            b'\x11\x22\x33\x44\x55\x66\x77\x88\x99\x00',
            b'\x11\x22\x33\x44\x55\x66\x77\x88\x00\x00',
            b'\x11\x22\x33\x44\x55\x66\x77\x00\x00\x00',
            b'\x11\x22\x33\x44\x55\x66\x00\x00\x00\x00',
            b'\x11\x22\x33\x44\x55\x00\x00\x00\x00\x00',
            b'\x11\x22\x33\x44\x00\x00\x00\x00\x00\x00',
            b'\x11\x22\x33\x00\x00\x00\x00\x00\x00\x00',
            b'\x11\x22\x00\x00\x00\x00\x00\x00\x00\x00',
            b'\x11\x00\x00\x00\x00\x00\x00\x00\x00\x00',
            b'\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00',
        ]

        for buf in bufs:
            form = self.Form.unpack(buf)
            assert form.data == dict(s=buf.rstrip(b'\x00'))


class TestStoreNumbersUpTo(unittest.TestCase):

    def test_unsigned_fields_correspond_to_numbers(self):
        mapping = [
            (0, minform.UInt8Field),
            (100, minform.UInt8Field),
            (255, minform.UInt8Field),
            (256, minform.UInt16Field),
            (500, minform.UInt16Field),
            (2**16 - 1, minform.UInt16Field),
            (2**16, minform.UInt32Field),
            (2**20, minform.UInt32Field),
            (2**32 - 1, minform.UInt32Field),
            (2**32, minform.UInt64Field),
            (2**40, minform.UInt64Field),
            (2**64 - 1, minform.UInt64Field),
        ]
        for n, typ in mapping:
            assert isinstance(minform.basic.store_numbers_up_to(n), typ)

    def test_signed_fields_correspond_to_numbers(self):
        mapping = [
            (0, minform.Int8Field),
            (100, minform.Int8Field),
            (127, minform.Int8Field),
            (128, minform.Int16Field),
            (500, minform.Int16Field),
            (2**15 - 1, minform.Int16Field),
            (2**15, minform.Int32Field),
            (2**20, minform.Int32Field),
            (2**31 - 1, minform.Int32Field),
            (2**31, minform.Int64Field),
            (2**40, minform.Int64Field),
            (2**63 - 1, minform.Int64Field),
        ]
        for n, typ in mapping:
            assert isinstance(minform.basic.store_numbers_up_to(n, True), typ)

    def test_unsigned_field_max(self):
        with pytest.raises(ValueError):
            minform.basic.store_numbers_up_to(2 ** 64)

    def test_signed_field_max(self):
        with pytest.raises(ValueError):
            minform.basic.store_numbers_up_to(2 ** 63, True)


minform/basic.py
import struct
import wtforms
from wtforms.validators import Length, NumberRange

from . import core


class BasicBinaryField(core.BinaryField):

    # Some BinaryFields will have inherent value restrictions, based on the
    # limitations of the serialized form. For example, a UInt8Field cannot
    # store numbers above 0xFF. When the class is instantiated, these
    # validators will be silently added to any validators provided by the
    # constructor.
    initial_validators = []

    def __init__(self, label='', validators=None, order=None, **kwargs):
        core.BinaryItem.__init__(self)
        self.size = struct.calcsize(self.pack_string)
        self.order = order

        # Clone the initial_validators list to avoid mutating a class
        # variable.
        all_vldtrs = list(self.initial_validators)
        if validators is not None:
            all_vldtrs.extend(validators)

        self.form_field = self.form_field_class(label, all_vldtrs, **kwargs)

    def pack(self, data, order=None):
        order = self.order or order or ''
        return self.pack_data(data, order)

    def pack_data(self, data, order):
        return struct.pack(order + self.pack_string, data)

    def unpack(self, buffer, order=None):
        order = self.order or order or ''
        return self.unpack_data(buffer, order)

    def unpack_data(self, buffer, order):
        return struct.unpack(order + self.pack_string, buffer)[0]


class CharField(BasicBinaryField):

    """
    Store a single byte as a one-character ``str`` (in Python 2) or ``bytes``
    object (in Python 3).

    Attributes:
        size: always ``1``
        form_field: A :class:`wtforms.fields.StringField` instance.
    """

    form_field_class = wtforms.StringField
    initial_validators = [Length(min=1, max=1)]
    pack_string = 'c'


class BinaryBooleanField(BasicBinaryField):

    """
    Store either ``True`` or ``False`` as ``b'\\x01'`` or ``b'\\x00'``
    (respectively).

    Attributes:
        size: always ``1``
        form_field: A :class:`wtforms.fields.BooleanField` instance.
    """

    form_field_class = wtforms.BooleanField
    pack_string = '?'


class BinaryIntegerField(BasicBinaryField):

    """
    This class should not be instantiated directly; instead, you should use
    one of its subclasses, which determine what kind of int is stored, and
    how. Those subclasses are:

    ==================== ==== =============== ================
    Name                 size Min             Max
    ==================== ==== =============== ================
    :class:`Int8Field`   1    -128            127
    :class:`UInt8Field`  1    0               255
    :class:`Int16Field`  2    -32768          32767
    :class:`UInt16Field` 2    0               65535
    :class:`Int32Field`  4    -2\ :sup:`31`   2\ :sup:`31` - 1
    :class:`UInt32Field` 4    0               2\ :sup:`32` - 1
    :class:`Int64Field`  8    -2\ :sup:`63`   2\ :sup:`63` - 1
    :class:`UInt64Field` 8    0               2\ :sup:`64` - 1
    ==================== ==== =============== ================

    Attributes:
        form_field: A :class:`wtforms.fields.Integerfield` instance.
    """

    form_field_class = wtforms.IntegerField

    @property
    def initial_validators(self):
        return [NumberRange(self.min, self.max)]


class Int8Field(BinaryIntegerField):

    pack_string = 'b'
    min = -128
    max = 127


class UInt8Field(BinaryIntegerField):

    pack_string = 'B'
    min = 0
    max = (2 ** 8) - 1


class Int16Field(BinaryIntegerField):

    pack_string = 'h'
    min = -(2 ** 15)
    max = (2 ** 15) - 1


class UInt16Field(BinaryIntegerField):

    pack_string = 'H'
    min = 0
    max = (2 ** 16) - 1


class Int32Field(BinaryIntegerField):

    pack_string = 'i'
    min = -(2 ** 31)
    max = (2 ** 31) - 1


class UInt32Field(BinaryIntegerField):

    pack_string = 'I'
    min = 0
    max = (2 ** 32) - 1


class Int64Field(BinaryIntegerField):

    pack_string = 'q'
    min = -(2 ** 63)
    max = (2 ** 63) - 1


class UInt64Field(BinaryIntegerField):

    pack_string = 'Q'
    min = 0
    max = (2 ** 64) - 1


class Float32Field(BasicBinaryField):

    """
    Store a ``float`` in four bytes.

    Attributes:
        size: Always ``4``.
        form_field: A :class:`wtforms.fields.FloatField` instance.
    """

    form_field_class = wtforms.FloatField
    pack_string = 'f'


class Float64Field(BasicBinaryField):

    """
    Store a ``float`` in eight bytes.

    Attributes:
        size: Always ``8``.
        form_field: A :class:`wtforms.fields.FloatField` instance.
    """

    form_field_class = wtforms.FloatField
    pack_string = 'd'


class BytesField(BasicBinaryField):

    """
    Store *N* bytes.

    Attributes:
        max_length: Maximum number of bytes in the stored string. Note that
            this may not be equal to :attr:`size`.

        size: The :attr:`size` of a :class:`BytesField` with ``max_length``
            *N* varies based on the *length* argument used to construct it.

            If *length* is :attr:`~minform.FIXED` or
            :attr:`~minform.AUTOMATIC`, ``size`` will be *N*.

            If *length* is :attr:`~minform.EXPLICIT`, there will be one or
            more extra bytes at the beginning of the packed data, which store
            the number of bytes used by the string. This will be the smallest
            number of bytes needed to store a number up to ``max_length``. So,
            ``size`` can be *N+1*, *N+2*, *N+4*, or *N+8*. (For more
            information, see the documentation for :data:`~minform.EXPLICIT`.)

        form_field: A :class:`wtforms.fields.StringField` instance.
    """

    form_field_class = wtforms.StringField

    def __init__(self, label='', validators=None, max_length=None,
                 length=core.AUTOMATIC, order=None, **kwargs):

        if not isinstance(max_length, int) or max_length < 0:
            raise ValueError('BytesField must be created with a '
                             'positive max_length keyword argument.')

        self.order = order
        self.length = length

        self.max_length = max_length

        if self.length == core.FIXED:
            self.initial_validators = [Length(max=max_length, min=max_length)]
            self.pack_string = '{0}s'.format(max_length)

        elif self.length == core.AUTOMATIC:
            self.initial_validators = [Length(max=max_length)]
            self.pack_string = '{0}s'.format(max_length)

        elif self.length == core.EXPLICIT:
            self.initial_validators = [Length(max=max_length)]
            self.length_field = store_numbers_up_to(max_length, order=order)
            self.pack_string = '{0}{1}s'.format(self.length_field.pack_string,
                                                max_length)

        super(BytesField, self).__init__(label, validators, order, **kwargs)

    def pack_data(self, data, order):
        buffer = bytearray(self.size)
        length = len(data)
        if self.length == core.EXPLICIT:
            pack_length_string = order + self.length_field.pack_string
            struct.pack_into(pack_length_string, buffer, 0, length)
            start = self.length_field.size
        else:
            start = 0
        buffer[start:start+length] = data
        return buffer

    def unpack_data(self, buffer, order):
        if self.length == core.EXPLICIT:
            unpack_length_string = order + self.length_field.pack_string
            length = struct.unpack_from(unpack_length_string, buffer)[0]
            if length > self.max_length:
                message = "Buffer cannot contain {0} bytes.".format(length)
                raise ValueError(message)
            data_buffer = buffer[self.length_field.size:]
        else:
            length = self.max_length
            data_buffer = buffer
        data = data_buffer[:length]

        if self.length == core.AUTOMATIC:
            data = data.rstrip(b'\x00')

        return data


def store_numbers_up_to(n, signed=False, **kwargs):
    """
    Return a BinaryField class that can store numbers up to a certain maximum.

    If the number is too big to store, a ``ValueError`` will be raised.

    Parameters:
        n: The highest number that you expect to need to store (must be at
            most a 64-bit integer).
        signed: Return a field that can store negative numbers.
        kwargs: Additional arguments get passed into the binary field
            constructor.

    Returns:
        BinaryIntegerField: A :class:`BinaryIntegerField` that can store
            numbers up to at least ``n``.
    """

    if signed:
        if n <= Int8Field.max:
            return Int8Field(**kwargs)
        elif n <= Int16Field.max:
            return Int16Field(**kwargs)
        elif n <= Int32Field.max:
            return Int32Field(**kwargs)
        elif n <= Int64Field.max:
            return Int64Field(**kwargs)
        else:
            raise ValueError("Can't track numbers up to {0}".format(n))
    else:
        if n <= UInt8Field.max:
            return UInt8Field(**kwargs)
        elif n <= UInt16Field.max:
            return UInt16Field(**kwargs)
        elif n <= UInt32Field.max:
            return UInt32Field(**kwargs)
        elif n <= UInt64Field.max:
            return UInt64Field(**kwargs)
        else:
            raise ValueError("Can't track numbers up to {0}".format(n))
